internal/tfimport/tfimport.go

```diff
@@ -560,7 +581,7 @@ func Run(rn runner.Runner, importRn runner.Runner, runArgs *RunArgs) error {
 // If it imported some resources but failed to import others, it will return true for retry. This is a simple way to solve dependencies without having to figure out the graph.
 // A specific case: GKE node pool name depends on random_id; import the random_id first, then do the cycle again and import the node pool.
 // skipped is a map to be filled with skipped resources so they are skipped on subsequent runs too.
-func planAndImport(rn, importRn runner.Runner, runArgs *RunArgs, skipped map[string]bool) (retry bool, err error) {
+func planAndImport(rn, importRn runner.Runner, runArgs *RunArgs, skipped map[string]bool) (successes, errs []string, err error) {
```

Comment:
why do we need both errs and err?

Reply:
I think of it as:
* errs is errors from terraform import
* err is an error in the operation of the importer itself that prevents it from continuing

Reply:
In that case consider creating a custom error type and return that.

You can use errors.As and fmt.Errorf with %w to check and wrap the error as needed

https://blog.golang.org/go1.13-errors

Reply:
I don't really want to treat errs as an error though. It's information for printing at the end, and may not get used at all if several retries ends up successfully importing all resources.

Reply:
How about returning a single error struct to differentiate between the two?

e.g.
type ImportError struct {
  Fatal  error
  Warnings []error
}

Reply:
It still doesn't seem right to me. It may be more accurate to call it failedImports rather than errs.

Reply:
In general:

- when you call a func and it returns nil error it means that everything was ok. Thus, I don't know if failedImports conveys that message
- if the caller wants to differentiate between different failures and handle some different than others, then idiomatic way to do is to return specific error types which can be handled by the caller independently.
- in cases where the caller does not care about differentiating then they can simply log/return the error

It's one of the reasons why go has errors as regular types instead of exceptions in languages like java etc, the caller is supposed to handle the error if there is something non trivial about it.

At the end of the day this is an internal func so consequences are not too high. If you want to improve it with different var naming + more documentation that's fine, but it would likely get flagged in a readability review, especially if a public func.

Reply:
Thank you, I've thought about this more and decide to do that, so I'm now using a custom error in the main retry loop.